src/background/index.ts

```diff
@@ -0,0 +1,68 @@
+import browser, { WebRequest } from 'webextension-polyfill'
+import type { Locale, MessageType } from '../types'
+
+declare const chrome: any
+let locale: Locale = null
+
+function updateLocale(value: Locale) {
+  if (value === locale) return
+  locale = value || null
+  browser.storage.local.set({ locale })
+  browser.browserAction.setBadgeText({ text: locale })
+}
+
+async function sendMessageToActiveTabs(message: MessageType) {
+  const active = await browser.tabs.query({ active: true, currentWindow: true })
+  for (const tab of active) {
+    if (tab.id) browser.tabs.sendMessage(tab.id, message)
+  }
+}
+
+async function getCurrentLocale() {
+  const active = await browser.tabs.query({ active: true, currentWindow: true })
+  for (const tab of active) {
+    if (tab.id) browser.tabs.sendMessage(tab.id, { type: 'get' })
+  }
+}
+
+// Intercept Accept-Language headers
+const options: WebRequest.OnBeforeSendHeadersOptions[] = ['blocking', 'requestHeaders']
+if (chrome.webRequest.OnBeforeSendHeadersOptions.hasOwnProperty('EXTRA_HEADERS')) {
+  options.push('extraHeaders')
+}
+browser.webRequest.onBeforeSendHeaders.addListener(
+  (details) => {
+    if (!locale) return
+    for (const header of details.requestHeaders || []) {
+      if (header.name.toLowerCase() === 'accept-language') {
+        header.value = locale
+        break
+      }
+    }
+    return { requestHeaders: details.requestHeaders || [] }
+  },
+  { urls: ['<all_urls>'] },
+  options
+)
+
+// Update on tab change or new urls
+browser.tabs.onActivated.addListener(() => getCurrentLocale())
```

Comment:
There's a subtle race condition here: when the active tab changes, we're asynchronously fetching the tab's locale. However, in the meantime the tab might have made network requests, and those will not pick up the right locale. I think this is also solved by maintaining a mapping from tabId to locale in the background script.

Reply:
Mhh the race condition should be fine, as that is only to pass it to the popup. The locale that is taken is saved locally.

Reply:
Nvm, I see what you mean. However the tab gets reloaded when it's changed. We will always have some kind of race condition as the commnunication is async.
Also generally it should be fine as the switch goes from: popup->bg->client.
Assuming you have multiple tabs, switchig and reloading faster that the message, but seems unlikely.